### PR TITLE
feat(misconf): Support `--ignore-policy` in config scans

### PR DESCRIPTION
**Pull request by simar7**
## Description

Adds support for passing `--ignore-policy` to ignore misconfiguration results in misconf scanning. Also correctly displays the exception count.

### Before
```shell
trivy  config  <input>

main.yaml (cloudformation)

Tests: 12 (SUCCESSES: 3, FAILURES: 9, EXCEPTIONS: 0)
Failures: 9 (UNKNOWN: 0, LOW: 2, MEDIUM: 1, HIGH: 6, CRITICAL: 0)
```

### After
```shell
trivy   config --ignore-policy=./config-ignore-policy.rego <input>

main.yaml (cloudformation)

Tests: 12 (SUCCESSES: 3, FAILURES: 8, EXCEPTIONS: 1)
Failures: 8 (UNKNOWN: 0, LOW: 2, MEDIUM: 1, HIGH: 5, CRITICAL: 0)

```

## Related issues
- Close https://github.com/aquasecurity/trivy/issues/3490

## Checklist
- [x] I've read the [guidelines for contributing](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/) to this repository.
- [x] I've followed the [conventions](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/#title) in the PR title.
- [x] I've added tests that prove my fix is effective or that my feature works.
- [ ] I've updated the [documentation](https://github.com/aquasecurity/trivy/blob/main/docs) with the relevant information (if needed).
- [ ] I've added usage information (if the PR introduces new options)
- [x] I've included a "before" and "after" example to the description (if the PR is a user interface change).

### CODE REVIEW

File: pkg/result/filter.go [hunk 1]
```diff
@@ -75,12 +75,36 @@ func FilterResult(ctx context.Context, result *types.Result, ignoreConf IgnoreCo
...
 
 	result.Vulnerabilities = filteredVulns
+	result.MisconfSummary = calcMisconfSummary(result, misconfSummary, filteredMisconfs)
 	result.Misconfigurations = filteredMisconfs
-	result.MisconfSummary = misconfSummary
 
 	return nil
 }
 
+func calcMisconfSummary(result *types.Result, misconfSummary *types.MisconfSummary, filteredMisconfs []types.DetectedMisconfiguration) *types.MisconfSummary {
```
Comment (knqyf263) [inline pkg/result/filter.go:84]: Can't we just increment exceptions here?
https://github.com/aquasecurity/trivy/blob/e40e1d6412ccc7be028f600e7818653949e898ff/pkg/result/filter.go#L163-L191

Comment (simar7) [reply pkg/result/filter.go:84]: We only know if the ignore policy applies or not to a misconfiguration once we make this call https://github.com/aquasecurity/trivy/blob/eea7d1558cf2ad81ee95bf315d580084bbea5684/pkg/result/filter.go#L273

Which is called here 
https://github.com/aquasecurity/trivy/blob/eea7d1558cf2ad81ee95bf315d580084bbea5684/pkg/result/filter.go#L70

File: pkg/flag/misconf_flags.go [hunk 2]
```diff
@@ -76,6 +76,7 @@ type MisconfFlagGroup struct {
 	PolicyBundleRepository *Flag
 
 	// Values Files
+	IgnorePolicy               *Flag
```
Comment (nikpivkin) [inline pkg/flag/misconf_flags.go:79]: `IgnorePolicy` is not assigned anywhere.

File: pkg/flag/options.go [hunk 3]
```diff
@@ -153,7 +153,7 @@ func (o *Options) FilterOpts() result.FilterOption {
 		IgnoreStatuses:     o.IgnoreStatuses,
 		IncludeNonFailures: o.IncludeNonFailures,
 		IgnoreFile:         o.IgnoreFile,
-		PolicyFile:         o.IgnorePolicy,
+		PolicyFile:         o.ReportOptions.IgnorePolicy,
```
Comment (nikpivkin) [inline pkg/flag/options.go:156]: should the field from `MisconfOptions` now be used?

File: pkg/result/filter.go [hunk 4]
```diff
@@ -75,12 +75,36 @@ func FilterResult(ctx context.Context, result *types.Result, ignoreConf IgnoreCo
...
+	if misconfSummary == nil {
+		return nil
+	}
+
+	var failed, passed int
```
Comment (nikpivkin) [inline pkg/result/filter.go:89]: `failed` is not used.

File: pkg/result/filter.go [hunk 5]
```diff
@@ -75,12 +75,36 @@ func FilterResult(ctx context.Context, result *types.Result, ignoreConf IgnoreCo
...
+		switch rm.Status {
+		case types.StatusPassed:
+			passed++
+		case types.StatusFailure:
+			failed++
+		}
+	}
+
+	if len(result.Misconfigurations)-len(filteredMisconfs) > 0 {
+		misconfSummary.Exceptions = len(result.Misconfigurations) - len(filteredMisconfs) - passed
```
Comment (nikpivkin) [inline pkg/result/filter.go:100]: Misconfiguration results are also filtered elsewhere, should we take them into account? https://github.com/aquasecurity/trivy/blob/e40e1d6412ccc7be028f600e7818653949e898ff/pkg/result/filter.go#L173

**Review by nikpivkin — CHANGES_REQUESTED** [review]
left a few comments

File: pkg/result/filter_test.go [hunk 6]
```diff
@@ -685,6 +685,66 @@ func TestFilter(t *testing.T) {
...
+				severities: []dbTypes.Severity{dbTypes.SeverityHigh},
+				policyFile: "./testdata/test-ignore-policy-misconf.rego",
+			},
+			want: types.Report{
+				Results: types.Results{
+					{
+						MisconfSummary: &types.MisconfSummary{
+							Successes:  1,
+							Failures:   2,
+							Exceptions: 1,
```
Comment (nikpivkin) [inline pkg/result/filter_test.go:732]: Shouldn't it be Failures = 0 and Exceptions = 2 here?

File: pkg/result/filter.go (hunk 1 above)
Comment (nikpivkin) [inline pkg/result/filter.go:84]: How about just getting the number of ignored misconfigurations by policy and adding to the result?

File: pkg/result/filter_test.go (hunk 6 above)
Comment (simar7) [reply pkg/result/filter_test.go:732]: Today we show failures that include exceptions. For example:

```shell
Tests: 13 (SUCCESSES: 3, FAILURES: 9, EXCEPTIONS: 1)
Failures: 8 (UNKNOWN: 0, LOW: 2, MEDIUM: 1, HIGH: 5, CRITICAL: 0)
```

In this case, 9 have failed but 1 of them is excepted. The remaining 8 are broken down into the next line by severity.

File: pkg/result/filter.go (hunk 1 above)
Comment (simar7) [reply pkg/result/filter.go:84]: could you expand on it?

File: pkg/result/filter_test.go (hunk 6 above)
Comment (nikpivkin) [reply pkg/result/filter_test.go:732]: @simar7 Ok, I didn't know that and didn't come across it in the documentation. But still, if two rules are ignored, then shouldn't there be 2 exceptions?

Comment (simar7) [reply pkg/result/filter_test.go:732]: 1 is a success and 2 are failures. Out of the 2 failures, one is excepted via  the exception rule (ignore policy). Hence 2 failures, 1 success and 1 exception.

File: pkg/result/filter.go (hunk 1 above)
Comment (knqyf263) [reply pkg/result/filter.go:84]: Oh, you're right. What if counting ignored results here?
https://github.com/aquasecurity/trivy/blob/d70b1ccd88f10730e8ca159a9a911de285a4faac/pkg/result/filter.go#L274-L276

Can't we return misconfSummary from `applyPolicy`?

File: pkg/result/filter.go [hunk 7]
```diff
@@ -168,6 +155,7 @@ func filterMisconfigurations(result *types.Result, severities []string, includeN
 			continue
 		} else if ignoreMisconfs.Match(result.Target, misconf.ID) || ignoreMisconfs.Match(result.Target, misconf.AVDID) {
 			// Filter misconfigurations by ignore file
+			ignored++
```
Comment (knqyf263) [inline pkg/result/filter.go:158]: I might be missing something. Why don't we increment `summary.Exceptions` here?

Comment (simar7) [reply pkg/result/filter.go]: You're right, I missed it. We can simplify this further.

File: pkg/result/filter.go (hunk 1 above)
Comment (simar7) [reply pkg/result/filter.go]: You're right I pushed some changes here https://github.com/aquasecurity/trivy/pull/5359/commits/4626583d726baf0fdada76b6a53edc48a6ab1252

**Review by knqyf263 — APPROVED** [review]

**Review by nikpivkin — APPROVED** [review]
lgtm

**Pull request merged**